### ISSUE TITLE
CKeditor wont render proccessed image if i set external link on image

### ISSUE THREAD

**Issue by agendartobias** [netresearch/t3x-rte_ckeditor_image#25]
Hello we have problem with rendering images in new CKeditor.

If we delete processed image from fileadmin/_processed_/.... we get 404 error in FE nad BE (doesent metter if we delete it over SSH or over install tool -> Clean up -> Clear processed files) - all filepermision are OK.
Image wont regenerate in FE and BE. If we go in problematic content and hit "Save" buttun again, image regenerate in BE and FE.

Can anyone help?

The problem only exsist if we set image some url over "link handler" (Exterlnal URL).

If we have only image with no clickable external url - image regenerate automatically.

**Comment by sypets** [comment]
It might be helpful to be able to trigger a regenerating of processed images (where the generated file does not exist yet).

Another issue: I have a number of images that are embedded with a path using "fileadmin/**_processed_**". It appears _processed_ is deprecated. These will automatically be fixed as well using a new path in fileadmin/_processed if the content element is opened and saved. It would be helpful to trigger doing this automatically as well.

**Comment by muh-nr** [comment]
@agendartobias 
Works for me on TYPO3 CMS 8.7.20.
Is this still a problem?

**Comment by agendartobias** [comment]
Yes, still the same problem. But i repeat it only exsist if we set image some url over "link handler" (like Exterlnal URL).

**Comment by Abdull** [comment]
My TYPO installation (`TYPO 9.5.8, rte_ckeditor_image  9.0.3`) is affected by this bug as well.

The bug is in `ImageLinkRenderingController.renderImages(..)`: This method currently does not leverage `MagicImageService` in order to generate a processed image if one doesn't exist currently (e.g., because _Remove Temporary Assets_ was triggered on `/fileadmin/_processed_/`).

Compare to `ImageRenderingController.renderImageAttributes(..)` which does use `MagicImageService` and successfully regenerates processed images on demand.

I incorporated `ImageRenderingController`'s logic into `ImageLinkRenderingController` and was able to successfully fix this issue:

    // inside ImageLinkRenderingController.php
    // ...
    public function renderImages($conf)
    {
        // Get link inner HTML
        $linkContent = $this->cObj->getCurrentVal();
        // Find all images with file-uid attribute
        $imgSearchPattern = '/<img(?=.*data-htmlarea-file-uid).*?\/>/';
        $attrSearchPattern = '/([a-zA-Z0-9-]+)=[\'"]([^\'"]*)/';
        $passedImages = [];
        $parsedImages = [];

        // Extract all TYPO3 images from link content
        preg_match_all($imgSearchPattern, $linkContent, $passedImages);
        $passedImages = $passedImages[0];

        if (count($passedImages) === 0) {
            return $linkContent;
        }

        foreach ($passedImages as $passedImage) {
            // Get image attributes
            preg_match_all($attrSearchPattern, $passedImage, $passedAttributes);
            $passedAttributes = array_combine($passedAttributes[1], $passedAttributes[2]);
            // Remove empty values
            $passedAttributes = array_filter($passedAttributes);

            if (!empty($passedAttributes['data-htmlarea-file-uid'])) {
                try {
                    $systemImage = Resource\ResourceFactory::getInstance()->getFileObject($passedAttributes['data-htmlarea-file-uid']);
                    $imageConfiguration = [
                        'width' => ($passedAttributes['width']) ? $passedAttributes['width'] : $systemImage->getProperty('width'),
                        'height' => ($passedAttributes['height']) ? $passedAttributes['height'] : $systemImage->getProperty('height')
                    ];
                    $processedFile = $this->getMagicImageService()->createMagicImage($systemImage, $imageConfiguration);
                    $imageAttributes = [
                        'src' => $processedFile->getPublicUrl(),
                        'title' => ($passedAttributes['title']) ? $passedAttributes['title'] : $systemImage->getProperty('title'),
                        'alt' => ($passedAttributes['alt']) ? $passedAttributes['alt'] : $systemImage->getProperty('alternative'),
                        'width' => ($passedAttributes['width']) ? $passedAttributes['width'] : $systemImage->getProperty('width'),
                        'height' => ($passedAttributes['height']) ? $passedAttributes['height'] : $systemImage->getProperty('height'),
                        'style' => ($passedAttributes['style']) ? $passedAttributes['style'] : $systemImage->getProperty('style')
                    ];
                    // Add original attributes, if not already parsed
                    $imageAttributes = $imageAttributes + $passedAttributes;
                    // Cleanup attributes; disable zoom images within links
                    $unsetParams = [
                        'data-htmlarea-file-uid',
                        'data-htmlarea-file-table',
                        'data-htmlarea-zoom',
                        'data-htmlarea-clickenlarge' // Legacy zoom property
                    ];
                    $imageAttributes = array_diff_key($imageAttributes, array_flip($unsetParams));
                    // Image template; empty attributes are removed by 3nd param 'false'
                    $parsedImages[] = '<img ' . GeneralUtility::implodeAttributes($imageAttributes, true, false) . ' />';
                } catch (Resource\Exception\FileDoesNotExistException $fileDoesNotExistException) {
                    $parsedImages[] = $passedImage;
                    // Log in fact the file could not be retrieved.
                    $message = sprintf('I could not find file with uid "%s"', $passedAttributes['data-htmlarea-file-uid']);
                    $this->getLogger()->error($message);
                }
            } else {
                $parsedImages[] = $passedImage;
            }
        }
        // Replace original images with parsed
        $linkContent = str_replace($passedImages, $parsedImages, $linkContent);

        return $linkContent;
    }

    /**
     * Instantiates and prepares the Magic Image service.
     *
     * @return \TYPO3\CMS\Core\Resource\Service\MagicImageService
     */
    protected function getMagicImageService()
    {
        /** @var $magicImageService Resource\Service\MagicImageService */
        static $magicImageService;
        if (!$magicImageService) {
            $magicImageService = GeneralUtility::makeInstance(\TYPO3\CMS\Core\Resource\Service\MagicImageService::class);
            // Get RTE configuration
            $pageTSConfig = $this->frontendController->getPagesTSconfig();
            if (is_array($pageTSConfig) && is_array($pageTSConfig['RTE.']['default.'])) {
                $magicImageService->setMagicImageMaximumDimensions($pageTSConfig['RTE.']['default.']);
            }
        }
        return $magicImageService;
    }
    
    // ...

It'd be great to have this bugfix soon in the next rte_ckeditor_image release.

**Comment by muh-nr** [comment]
@Abdull 
Thanks for investigating this. I totally missed this point up...

Please check current master:
The processed images should always be regenerated after viewing FE or saving BE.

**Comment by Abdull** [comment]
@muh-nr, thank you very much! I have _composer-required_ `"netresearch/rte-ckeditor-image": "dev-master#7ad1bdd622f77aa338e35365d80e56acb54f44b3"` and it works like a charm.

**Comment by muh-nr** [comment]
Benissimo!
